pkg/chunk/purger/purger.go

```diff
@@ -0,0 +1,580 @@
+)
+
+const millisecondPerDay = int64(24 * time.Hour / time.Millisecond)
+
+var separatorString = string([]byte{model.SeparatorByte})
+
+type deleteRequestWithLogger struct {
+	chunk.DeleteRequest
+	logger log.Logger
+}
+
+// DataPurgerConfig holds config for DataPurger
+type DataPurgerConfig struct {
+	EnablePurger    bool   `yaml:"enable_purger"`
+	NumWorkers      int    `yaml:"num_workers"`
+	ObjectStoreType string `yaml:"object_store_type"`
+}
+
+// RegisterFlags registers CLI flags for DataPurgerConfig
+func (cfg *DataPurgerConfig) RegisterFlags(f *flag.FlagSet) {
+	f.BoolVar(&cfg.EnablePurger, "purger.enable-purger", false, "Enable purger to allow deletion of series. Be aware that Delete series feature is still experimental")
+	f.IntVar(&cfg.NumWorkers, "purger.num-workers", 2, "Number of workers executing delete plans in parallel")
+	f.StringVar(&cfg.ObjectStoreType, "purger.object-store-type", "", "Name of the object store to use for storing delete plans")
+}
+
+type workerJob struct {
+	planNo          int
+	userID          string
+	deleteRequestID string
+	logger          log.Logger
+}
+
+type workerJobExecutionStatus struct {
+	workerJob
+	err error
+}
+
+// DataPurger does the purging of data which is requested to be deleted
+type DataPurger struct {
+	cfg           DataPurgerConfig
+	deleteStore   *chunk.DeleteStore
+	chunkStore    chunk.Store
+	storageClient chunk.ObjectClient
+
+	executePlansChan             chan deleteRequestWithLogger
+	workerJobChan                chan workerJob
+	workerJobExecutionStatusChan chan workerJobExecutionStatus
+
+	// we would only allow processing of singe delete request at a time since delete requests touching same chunks could change the chunk IDs of partially deleted chunks
+	// and break the purge plan for other requests
+	inProcessRequestIDs    map[string]string
+	inProcessRequestIDsMtx sync.RWMutex
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// NewDataPurger creates a new DataPurger
+func NewDataPurger(cfg DataPurgerConfig, deleteStore *chunk.DeleteStore, chunkStore chunk.Store, storageClient chunk.ObjectClient) (*DataPurger, error) {
+	dataPurger := DataPurger{
+		cfg:                          cfg,
+		deleteStore:                  deleteStore,
+		chunkStore:                   chunkStore,
+		storageClient:                storageClient,
+		inProcessRequestIDs:          map[string]string{},
+		executePlansChan:             make(chan deleteRequestWithLogger, 50),
+		workerJobChan:                make(chan workerJob, 50),
+		workerJobExecutionStatusChan: make(chan workerJobExecutionStatus, 50),
+		quit:                         make(chan struct{}),
+	}
+
+	dataPurger.runWorkers()
+	go dataPurger.jobScheduler()
+
+	err := dataPurger.loadInprocessDeleteRequests()
+	if err != nil {
+		return nil, err
+	}
+
+	return &dataPurger, nil
+}
+
+// Run starts workers, job scheduler and a loop for feeding delete requests to them for processing
+func (dp *DataPurger) Run() {
+	dp.loop()
```

Comment:
We can probably get rid of the `loop` function and just move the code from that function here. I don't think the loop function is called anywhere else